packages/gateway/src/features/registration/type-resolvers.ts

```diff
@@ -721,7 +722,31 @@ export const typeResolvers: GQLResolver = {
     id(docRef: fhir.DocumentReference) {
       return (docRef.masterIdentifier && docRef.masterIdentifier.value) || null
     },
-    data(docRef: fhir.DocumentReference) {
+    data: async (docRef: fhir.DocumentReference, _, authHeader) => {
```

Comment:
This should just be a URL reference to the GET /document handler, right? If we embed the base64 here, the browser cannot cache this data and I'm pretty sure we again make these queries a lot slower. With large attachment my concern is also that we quickly run out of space in the device storage..

I think someone having access to one file with a random URL is not big enough a security problem for now to make such a big compromise in complexity and performance.

Reply:
@rikukissa About the queries being slower, well even with using the minio urls instead of base64 value, as we needed to support offline access to the attachments, in client we needed to "fetch" the attachments as part of the download and then cache them in workbox. So I'm not sure if it would affect the download speed, rather it might be faster as we are downloading them from gateway which is closer to our minio server no?

Reply:
@rikukissa essentially the concern we have is that just as with the QR code on verification of a certificate, it will be possible for an attachment URL to be shared.  We took the decision that was not worth any reputational risk.  It would not make sense to take a differing view on attachments. Therefore I feel we have to compromise here.

Reply:
@euanmillar 

I think Pyry’s recommended approach discussed on Slack was a solution where we wouldn’t have to make a compromise in performance, low-connectivity support or in added complexity. So in summary:

- Create temporary presigned URLs to attachments in gateway that are only valid for 30 minutes (or some other configurable time). If the URL is leaked, it’s only available for 30 minutes. We also can easily identify which user leaked it.
- Client can now download the attachments just as with any urls and store the files for offline usage
- No code changes needed anywhere in the client – less chance of us breaking or worsening these workflows.

**The problems with reading files to base64:**

- None of the attachments can now be cached by the browser
- One record could have many megabytes of data. A record doesn’t need many attachments before the payload to download would be more than the OpenCRVS client itself. Users using a flaky or low-bandwidth connection might never be able to download records because of this. I think our goal should be to minimise the amount of data that the client needs to download.
    - Also note that base64 is not meant to be sent across the wire. It’s one of the worst ways of transmitting files filesize wise. Base64 encoded filesize will always be more than the original file.
- To be able to attach the Base64 file to the GraphQL handler response the file needs to be read to the server’s memory first. Luckily our files are not hundreds of megs, but making this operation concurrently when multiple users are downloading records can cause us running low on memory. What if one day we want to support PDFs?

My recommendation is we start with Pyry’s approach as it’s the industry best practice of handling this situation and only then if it doesn’t work we make our custom solution.

Reply:
https://github.com/minio/minio-js/blob/master/examples/presigned-getobject.js

Reply:
Yeah sounds good. @sadmananik Could you make the necessary changes please? Apologies but we will have to change the approach. I say let's keep this PR as it is now and make another PR with the `presigned` url approach